pkg/util/checkpoint.go

```diff
@@ -0,0 +1,319 @@
+	// Hash is the hash which commits to the contents of the entire log.
+	Hash []byte
+	// OtherContent is any additional data to be included in the signed payload; each element is assumed to be one line
+	OtherContent []string
+}
+
+// String returns the String representation of the Checkpoint
+func (c Checkpoint) String() string {
+	var b strings.Builder
+	fmt.Fprintf(&b, "%s\n%d\n%s\n", c.Ecosystem, c.Size, base64.StdEncoding.EncodeToString(c.Hash))
+	for _, line := range c.OtherContent {
+		fmt.Fprintf(&b, "%s\n", line)
+	}
+	return b.String()
+}
+
+// MarshalText returns the common format representation of this Checkpoint.
+func (c Checkpoint) MarshalText() ([]byte, error) {
+	return []byte(c.String()), nil
+}
+
+// UnmarshalText parses the common formatted checkpoint data and stores the result
+// in the Checkpoint.
+//
+// The supplied data is expected to begin with the following 3 lines of text,
+// each followed by a newline:
+// <ecosystem/version string>
+// <decimal representation of log size>
+// <base64 representation of root hash>
+// <optional non-empty line of other content>...
+// <optional non-empty line of other content>...
+//
+// This will discard any content found after the checkpoint (including signatures)
+func (c *Checkpoint) UnmarshalText(data []byte) error {
+	l := bytes.Split(data, []byte("\n"))
+	if len(l) < 4 {
+		return errors.New("invalid checkpoint - too few newlines")
+	}
+	eco := string(l[0])
+	if len(eco) == 0 {
+		return errors.New("invalid checkpoint - empty ecosystem")
+	}
+	size, err := strconv.ParseUint(string(l[1]), 10, 64)
+	if err != nil {
+		return fmt.Errorf("invalid checkpoint - size invalid: %w", err)
+	}
+	h, err := base64.StdEncoding.DecodeString(string(l[2]))
+	if err != nil {
+		return fmt.Errorf("invalid checkpoint - invalid hash: %w", err)
+	}
+	*c = Checkpoint{
+		Ecosystem: eco,
+		Size:      size,
+		Hash:      h,
+	}
+	if len(l) >= 5 {
+		for _, line := range l[3:] {
+			if len(line) == 0 {
+				break
+			}
+			c.OtherContent = append(c.OtherContent, string(line))
+		}
+	}
+	return nil
+}
+
+func (c Checkpoint) Sign(identity string, signer crypto.Signer, opts crypto.SignerOpts) (*note.Signature, error) {
+	hf := crypto.SHA256
+	if opts != nil {
+		hf = opts.HashFunc()
+	}
+
+	input, _ := c.MarshalText()
+	var digest []byte
+	if hf != crypto.Hash(0) {
+		hasher := hf.New()
+		_, err := hasher.Write(input)
+		if err != nil {
+			return nil, errors.Wrap(err, "hashing checkpoint before signing")
+		}
+		digest = hasher.Sum(nil)
+	} else {
+		digest, _ = c.MarshalText()
+	}
+
+	sig, err := signer.Sign(rand.Reader, digest, opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "signing checkpoint")
+	}
+	pubKeyBytes, err := x509.MarshalPKIXPublicKey(signer.Public())
+	if err != nil {
+		return nil, errors.Wrap(err, "marshalling public key")
+	}
+
+	pkSha := sha256.Sum256(pubKeyBytes)
+
+	signature := note.Signature{
+		Name:   identity,
+		Hash:   binary.BigEndian.Uint32(pkSha[:]),
```

Comment:
is it important that the algorithm is encoded in this hash? 
it is included in the sumdb note code
https://cs.opensource.google/go/x/mod/+/refs/tags/v0.4.2:sumdb/note/note.go;drc=ed3ec21bb8e252814c380df79a80f366440ddb2d;l=326